src/components/AddProductDialog.vue

```diff
@@ -334,15 +341,67 @@ function transformVariant(
             taxRateId: variant.taxRateId,
             weight:
                 variant.weight !== undefined ? parseFloat(variant.weight) : 0,
+            mediaIds: variant.mediaIds,
         },
         isPubliclyVisible: !variant.invisible,
     }
 }
 
+/**
+ * Uploads files to the media service via GraphQL.
+ * @param files Files to upload.
+ */
+async function uploadMedias(files: File[]) {
+    const uploadPromises = files.map(uploadFileWorkaround)
+    const mediaIds = await Promise.all(uploadPromises)
+    return mediaIds
+}
+
+/**
+ * Uploads file to media service via GraphQL.
+ * Uses manually constructed request as library dropped file-upload support.
+ * We do not want to use express.
+ * @param file The file to upload.
+ */
+async function uploadFileWorkaround(file: File) {
+    const queryAndVariables = {
+        query: `mutation ($file: Upload!) {
+          uploadMedia (mediaFile: $file)
+        }`,
+        variables: {
+            file: null,
+        },
+    }
+    const map = {
+        '0': ['variables.file'],
+    }
+    let formData = new FormData()
+    formData.append('operations', JSON.stringify(queryAndVariables))
+    formData.append('map', JSON.stringify(map))
+    formData.append('0', file)
+    const token = `Bearer ${await store.getAccessToken(true)}`
+    const requestOptions = {
+        method: 'POST',
+        headers: { Authorization: token },
+        body: formData,
+    }
+    const data = await fetch('/api/graphql', requestOptions).then(
+        (response) => {
+            return response.json()
+        }
+    )
+    return data.data.uploadMedia
+}
+
 /**
  * Tries to save the product and its variants (to the catalog service).
  */
 async function save() {
+    for (const variant of variants.value) {
```

Comment:
But the user might not know that it is the file upload that is causing the product addition to fail. Because the error message is too general. It just says that something went wrong. Could you catch the error of the file upload and display an error message telling the user that the file upload is causing problems? Then the user could make the decision to skip the file upload. `errorHandler.ts` has a function to push error notifications: `pushErrorNotification`.

Reply:
Yes, this is implemented in the review changes with `awaitActionAndPushErrorIfNecessary`